### ISSUE TITLE
volume controller not handling terminated node

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#31088]
When I terminated a node with volumes attached, the volume was not detached, because the volume controller was still looking for the instance:

```
I0822 02:15:26.521662       6 nodecontroller.go:586] Deleting node (no longer present in cloud provider): ip-172-20-116-14.ec2.internal
I0822 02:15:26.521700       6 nodecontroller.go:663] Recording Deleting Node ip-172-20-116-14.ec2.internal because it's not present according to cloud provider event message for node ip-172-20-116-14.ec2.internal
I0822 02:15:26.522162       6 event.go:216] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"ip-172-20-116-14.ec2.internal", UID:"ip-172-20-116-14.ec2.internal", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'DeletingNode' Node ip-172-20-116-14.ec2.internal event: Deleting Node ip-172-20-116-14.ec2.internal because it's not present according to cloud provider
I0822 02:15:26.530355       6 replica_set.go:446] Too few "default"/"b-6233150" replicas, need 1, creating 1
I0822 02:15:26.540855       6 event.go:216] Event(api.ObjectReference{Kind:"ReplicaSet", Namespace:"default", Name:"b-6233150", UID:"ebd55faa-680b-11e6-9e4b-0ad778e2c547", APIVersion:"extensions", ResourceVersion:"5356", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: b-6233150-z42jz
I0822 02:15:26.546594       6 replication_controller.go:480] Too few "kube-system"/"kube-dns-v14" replicas, need 2, creating 1
I0822 02:15:26.584561       6 event.go:216] Event(api.ObjectReference{Kind:"ReplicationController", Namespace:"kube-system", Name:"kube-dns-v14", UID:"372451bc-6805-11e6-ae22-0ad778e2c547", APIVersion:"v1", ResourceVersion:"54", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: kube-dns-v14-z1yxb
E0822 02:15:26.594877       6 deployment_controller.go:400] Error syncing deployment default/b: Operation cannot be fulfilled on deployments.extensions "b": the object has been modified; please apply your changes to the latest version and try again
I0822 02:15:27.156400       6 reconciler.go:115] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-1c140fb1" from node "ip-172-20-116-14.ec2.internal"
E0822 02:15:27.157735       6 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/vol-1c140fb1\"" failed. No retries permitted until 2016-08-22 02:17:27.157707473 +0000 UTC (durationBeforeRetry 2m0s). Error: DetachVolume failed fetching node from API server for volume "kubernetes.io/aws-ebs/vol-1c140fb1" (spec.Name: "mypd") from node "ip-172-20-116-14.ec2.internal" with: nodes "ip-172-20-116-14.ec2.internal" not found
I0822 02:15:31.524061       6 nodecontroller.go:512] NodeController observed a Node deletion: ip-172-20-116-14.ec2.internal
I0822 02:15:31.524086       6 nodecontroller.go:663] Recording Removing Node ip-172-20-116-14.ec2.internal from NodeController event message for node ip-172-20-116-14.ec2.internal
I0822 02:15:31.524273       6 event.go:216] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"ip-172-20-116-14.ec2.internal", UID:"ip-172-20-116-14.ec2.internal", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RemovingNode' Node ip-172-20-116-14.ec2.internal event: Removing Node ip-172-20-116-14.ec2.internal from NodeController
I0822 02:15:33.728782       6 servicecontroller.go:685] Detected change in list of current cluster nodes. New node set: [ip-172-20-116-13.ec2.internal]
I0822 02:15:33.728849       6 servicecontroller.go:693] Successfully updated 2 out of 2 load balancers to direct traffic to the updated set of nodes
W0822 02:16:15.734725       6 reflector.go:334] pkg/controller/service/servicecontroller.go:150: watch of *api.Service ended with: couldn't queue object: couldn't create key for object &{lastState:0xc820aa8d00 appliedState:0xc820aa8d00 mu:{state:0 sema:0} lastRetryDelay:0}: object has no meta: struct service.cachedService lacks embedded TypeMeta type
I0822 02:16:16.735996       6 servicecontroller.go:243] Got new Sync delta for service: default/kubernetes
I0822 02:16:16.737550       6 servicecontroller.go:305] LB doesn't need update for service default/kubernetes
I0822 02:16:16.737583       6 servicecontroller.go:243] Got new Sync delta for service: kube-system/kube-dns
I0822 02:16:16.738606       6 servicecontroller.go:305] LB doesn't need update for service kube-system/kube-dns
I0822 02:17:27.252838       6 reconciler.go:115] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-1c140fb1" from node "ip-172-20-116-14.ec2.internal"
E0822 02:17:27.254320       6 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/vol-1c140fb1\"" failed. No retries permitted until 2016-08-22 02:19:27.254287127 +0000 UTC (durationBeforeRetry 2m0s). Error: DetachVolume failed fetching node from API server for volume "kubernetes.io/aws-ebs/vol-1c140fb1" (spec.Name: "mypd") from node "ip-172-20-116-14.ec2.internal" with: nodes "ip-172-20-116-14.ec2.internal" not found
```

and that repeats forever (?):

```
I0822 02:21:27.416614       6 reconciler.go:115] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-1c140fb1" from node "ip-172-20-116-14.ec2.internal"
E0822 02:21:27.417646       6 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/vol-1c140fb1\"" failed. No retries permitted until 2016-08-22 02:23:27.417620039 +0000 UTC (durationBeforeRetry 2m0s). Error: DetachVolume failed fetching node from API server for volume "kubernetes.io/aws-ebs/vol-1c140fb1" (spec.Name: "mypd") from node "ip-172-20-116-14.ec2.internal" with: nodes "ip-172-20-116-14.ec2.internal" not found
I0822 02:23:27.496254       6 reconciler.go:115] Started DetachVolume for volume "kubernetes.io/aws-ebs/vol-1c140fb1" from node "ip-172-20-116-14.ec2.internal"
E0822 02:23:27.497630       6 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/vol-1c140fb1\"" failed. No retries permitted until 2016-08-22 02:25:27.497605095 +0000 UTC (durationBeforeRetry 2m0s). Error: DetachVolume failed fetching node from API server for volume "kubernetes.io/aws-ebs/vol-1c140fb1" (spec.Name: "mypd") from node "ip-172-20-116-14.ec2.internal" with: nodes "ip-172-20-116-14.ec2.internal" not found
```

**Comment by justinsb** [comment]
cc @saad-ali

**Comment by justinsb** [comment]
```
Server Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.5", GitCommit:"b0deb2eb8f4037421077f77cb163dbb4c0a2a9f5", GitTreeState:"clean", BuildDate:"2016-08-11T20:21:58Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by saad-ali** [comment]
This looks like a dupe of https://github.com/kubernetes/kubernetes/issues/29358 which was fixed by PRs https://github.com/kubernetes/kubernetes/issues/29485 and https://github.com/kubernetes/kubernetes/pull/30737 and should be in v1.3.6.

I'll close this issue for now. Please reopen if you run into the issue on v1.3.6 or later.

**Comment by justinsb** [comment]
SGTM @saad-ali ... when we get a fix for #29324 remind me to validate by terminating the instances :-)
